pkg/bigquery/sanitize.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bigquery implements helpers for using BigQuery.
+package bigquery
+
+import (
+	"fmt"
+	"regexp"
+	"unicode/utf8"
+)
+
+const tableNameMaxUTF8Bytes = 1024
+
+// Start with lowercase, middle is lowercase, number or hyphen, cannot end in
+// hyphen. 6-30 characters in length (start, 4-28 middle, end).
+var projectIDMatcher = regexp.MustCompile(`^[a-z][a-z0-9\-]{4,28}[a-z0-9]$`)
+
+// Lowercase and uppercase letters and underscores. Max 1024 characters.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const datasetIDRegex = `^[a-zA-Z0-9_]{1,512}[a-zA-Z0-9_]{0,512}$`
+
+var datasetIDMatcher = regexp.MustCompile(datasetIDRegex)
+
+// Unicode characters in category L (letter), M (mark), N (number),
+// Pc (connector, including underscore), Pd (dash), Zs (space). Max 1024
+// characters. 1024 is actually UTF-8 bytes from experimentation.
+// UTF-8 length check will be done separately.
+// regexp only allows 1000 repetitions, so had to manually repeat.
+const tableNameRegex = `^[\p{L}\p{M}\p{N}\p{Pc}\p{Pd}\p{Zs}]+$`
+
+var tableNameMatcher = regexp.MustCompile(tableNameRegex)
```

Comment:
These all seem to be focused on validation instead of extraction? I'm generally -1 on overzealous client-side validation, because it puts us in a precarious position when the server changes constraints. In general, we should let the server fail and give the user the appropriate response.

As written, if GCP allowed 30-character project IDs, our tool would have to change. I think that's a poor user experience. Same for the rest of these things. Client-side validation is rarely helpful.

I'm also not seeing how these regexes are protecting against injection attacks.

Reply:
My assumption is that valid table names cannot produce an injection attack as we can be sure that there are no non-identifier characters. If that was possible, there would be some serious problems with table naming rules.

Another option would be to look for cases of "bad" strings such as `";` but disallowing all possible bad strings is a harder problem than only allowing known good strings.

Reply:
Though maybe just checking for no "`" chars is sufficient. I wasn't sure if there was some clever unicode trick or something that could get around it.

Reply:
I would probably make a reasonably positioned regular expression here, match against the table name, log a warning if it doesn't match, and then continue with processing.